man/man1/feedwarrior.1
.TH feedwarrior 1


.SH NAME
feedwarrior - Feeds, warrior style


.SH DESCRIPTION

Inspired by
.UR https\://taskwarrior.org
.UE ,
this project seeks to simplify the maintenance of engineering work logs.

It maintains a collection of logs.

Entries of each individual log has core text content. Additionally, every entry may have any number of file attachments associated with it.


.SH SYNOPSIS

.SY feedwarrior
add [ -c config_file] [ -z ] [ --task-id ] [ --task-uuid ] < -s subject >
.YS

.SY feedwarrior
attach [ -c config_file] [ -z ] < -e entry > file
.YS

.SY feedwarrior
create [ -c config_file] [ --alias name ]
.YS

.SY feedwarrior
entry [ -c config_file] [ -z ] [ --task-id ] [ --task-uuid ] < -s subject > < path_to_content >
.YS

.SY feedwarrior
list [ -c config_file]
.YS

.SY feedwarrior
show [ -c config_file] < -l feed_id >
.YS


.SS Common options

.TP
.BI -l log_id
Log id to operate on. May be the full uuid or the alias defined at creation time.

.TP
.BI -c config_file
Load configuration from file.

.SS Options for show

.TP
.B --headers
Include multipart-headers in output.

.SS Options for attach

.TP
.BI -e entry_id
uuid of entry to attach file to.

.SS Options for create

.TP
.BI --alias alias_id
Create log id alias that may be used as argument to \fB-l\fP for future operations.


.SS Options for add and entry

.TP
.BI -s subject
Entry subject.

.TP
.BI --task-id id
Associate the log entry with the given taskwarrior numeric id. This is a quasi-magical instruction which will resolve to the actual taskwarrior uuid.

.TP
.BI --task-uuid uuid
Associate the log entry with the given taskwarrior numeric uuid.


.SH ATTACHMENTS

Both \fBfeedwarrior add\fP and \fBfeedwarrior entry\fP will return the created entry id.

This id can be used further with the \fB-e\fP flag for attachments.

Any file in any format may be an attachment. The feedwarrior tool imposes little if no restriction on what data qualifies as valid attachment data.


.SH INTERACTIVE ENTRIES

The \fBfeedwarrior add\fP command enables interactive definition of textual update entry contents.

It will open an editor (at the moment hardcoded to \fBvim\fP) to receive the contents.


.SH EXAMPLE

.EX
feedwarrior create --alias myfeed
echo "this is some content" > /tmp/foo.txt
entry=`feedwarrior entry -s "first update" /tmp/foo.txt`
feedwarrior attach -e $entry <attachment_file>
.EE


.SH LICENSE

This documentation and its source is licensed under the Creative Commons Attribution-Sharealike 4.0 International license.

The source code of the tool this documentation describes is licensed under the GNU General Public License 3.0.


.SH COPYRIGHT AND CONTACT

.MT dev@holbrook.no
Louis Holbrook
.ME

.UR https://holbrook.no
.UE https://holbrook.no

.P
PGP: 59A844A484AC11253D3A3E9DCDCBD24DD1D0E001


.SH SOURCE CODE

https://holbrook.no/src/feedwarrior/log.htm ggj


CHANGELOG
* 0.5.6
	- Add long description for pypi 
* 0.5.5
	- Add man page
	- Return entry uuid on add and entry subcommands
* 0.5.4
	- Package for pypi.org
* 0.5.3
	- Package
* 0.5.2
	- Alert when running with missing configuration
* 0.5.1
	- Fix sort on date
* 0.5.0 
	- Add taskwarrior uuid lookup and id to uuid resolve
	- Abort add on empty input
* 0.4.2 
	- Add "add" subcommand
* 0.4.1 
	- Add list feeds subcommand
* 0.4.0
	- Consolidate put and get methods on same adapter interface
	- Resolve uncompressed path bug 
* 0.3.0 
	- Add gzip option for entries
* 0.2.2
	- Add all inline text content to output in show cmd
* 0.2.1
	- Fix non-committed headers flag
* 0.2.0
	- Rename raw flag to headers
* 0.1.4 
	- Add flag for raw headers in entry output
* 0.1.3
	- Add multiple hashers for each entry 
	- Fix bug in config handler that prevents handling of config filenames passed to it
	- Fix opinionated arg flags before or after subcommand
* 0.1.2
	- Add simple, readable feed summary
	- Correct binary script name in setup.py
* 0.1.1
	- Add feed alias on creation
* 0.1.0
	- Add feed, entry, ui framework and setup


feedwarrior.py
#!/usr/bin/python

# Author: Louis Holbrook <dev@holbrook.no> (https://holbrook.no)
# License: GPLv3
# Description: Work log tool

# standard imports
import copy
import os
import sys
import argparse
import configparser
import json
import logging
import uuid
import time

logging.basicConfig(level=logging.ERROR)
logg = logging.getLogger()

DATA_DIR='/dev/null'


argparser = argparse.ArgumentParser(description='create and manipulate feedwarrior logs')
argparser.add_argument('-p', type=str, help='parent log uuid')
argparser.add_argument('-c', required=True, type=str, help='configuration file')
argparser.add_argument('-v', action='store_true', help='be verbose')
sub = argparser.add_subparsers()
sub.dest = 'command'

args = argparser.parse_args(sys.argv[1:])
if args.v:
    logging.getLogger().setLevel(logging.DEBUG)

logg.debug('loading config {}'.format(args.c))
cp = configparser.ConfigParser()
cp.read(args.c)
DATA_DIR = cp['FEEDWARRIOR']['datadir']

try:
    os.mkdir(DATA_DIR)
    logg.debug('creating datadir {}'.format(DATA_DIR))
except FileExistsError as e:
    logg.debug('using datadir {}'.format(DATA_DIR))


# TODO: move to submodule asap

def parse_uuid(uu):
    if type(uu).__name__ == 'str':
        return uuid.UUID('urn:uuid:' + uu)
    elif type(uu).__name__ == 'UUID':
        return uu
    raise ValueError('invalid uuid')

class Feed:

    def __init__(self, uu=None, parent=None, created=None, updated=None):
        if uu == None:
            self.uuid = uuid.uuid4()
        else:
            self.uuid = parse_uuid(uu)

        self.parent = None
        if parent != None:
            if type(parent).__name__ != 'Feed':
                raise ValueError('wrong type for parent: {}'.format(type(parent).__name__))
            self.parent = parent

        if created != None:
            self.created = created
            if updated == None:
                self.updated = copy.copy(created)
        else:
            self.created = int(time.time())
            self.updated = copy.copy(self.created)

        if self.updated == None:
            self.updated = updated

        self.entries = []


    def serialize(self):
        o = {
            'uuid': str(self.uuid),
            'created': self.created,
            'updated': self.updated,
                }
        if self.parent != None:
            o['parent_uuid'] = str(self.parent.uuid)
        
        return o

feed_parent = None
if args.p != None:
    try:
        feed_parent = Feed(args.p)
    except ValueError as e:
        logg.error('invalid parent {}: {}'.format(args.p, e))
        sys.exit(1)

if args.command == None:
    feed_current = Feed(parent=feed_parent)
    uu = str(feed_current.uuid)
    logg.debug('new log {}'.format(uu))
    log_path = os.path.join(DATA_DIR, str(uu))
    os.mkdir(log_path)

    log_meta_path = os.path.join(log_path, '.log')
    f = open(log_meta_path, 'x')
    json.dump(feed_current.serialize(), f)
    f.close()
    sys.exit(0)
